platform/ios/src/MGLMapView.mm

```diff
@@ -3139,18 +3139,30 @@ - (BOOL)isRotationAllowed
     return (self.zoomLevel >= self.currentMinimumZoom);
 }
 
-// correct rotations to north as needed
-//
-- (void)unrotateIfNeededAnimated:(BOOL)animated
+- (void)unrotateIfNeededForGesture
 {
-    double snapTolerance = 7;
-
-    // don't worry about it in the midst of pinch or rotate gestures
-    //
-    if (self.pinch.state  == UIGestureRecognizerStateChanged || self.rotate.state == UIGestureRecognizerStateChanged) return;
+    // Avoid contention with in-progress gestures.
+    UIGestureRecognizerState state = self.pinch.state;
+    if (self.direction != 0
+        && state != UIGestureRecognizerStateBegan
+        && state != UIGestureRecognizerStateChanged)
+    {
+        [self unrotateIfNeededAnimated:YES];
+        
+        // Snap to north.
+        if ((self.direction < MGLToleranceForSnappingToNorth
+             || self.direction > 360 - MGLToleranceForSnappingToNorth)
+            && self.userTrackingMode != MGLUserTrackingModeFollowWithHeading
+            && self.userTrackingMode != MGLUserTrackingModeFollowWithCourse)
```

Comment:
Good catch. I was initially unsure about needing to check for the tracking mode here, but two-finger zoom and quickzoom both require it.